### PR TITLE
[JENKINS-55582] Split instance-identity to a plugin

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/instance-identity-plugin/pull/17. ~Requires #5049 because `sshd` depends on `instance-identity` (for real).~

### Proposed changelog entries

* Converting the `instance-identity` module into a (detached) plugin.

### Proposed upgrade guidelines

Normally the new plugin will be automatically installed. If you use an “as-code” installation mechanism like a text file with a list of plugins, you may need to add `instance-identity` (version `3.0` or later) in order to restore this functionality.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/jenkins/split-plugins.txt [hunk 1]
```diff
@@ -20,7 +20,7 @@ windows-slaves 1.547 1.0
 antisamy-markup-formatter 1.553 1.0
 matrix-project 1.561 1.0
 junit 1.577 1.0
-bouncycastle-api 2.16 2.16.0
+bouncycastle-api 2.18 2.16.0
```
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt:23]: Dep of new plugin.

File: test/src/test/java/jenkins/security/ClassFilterImplTest.java [hunk 2]
```diff
@@ -159,7 +159,6 @@ public void moduleClassesShouldBeWhitelisted() throws Exception {
         ClassFilterImpl filter = new ClassFilterImpl();
         filter.check("org.jenkinsci.main.modules.cli.auth.ssh.UserPropertyImpl");
         filter.check("org.jenkinsci.modules.windows_slave_installer.WindowsSlaveInstaller");
-        filter.check("org.jenkinsci.main.modules.instance_identity.PageDecoratorImpl");
```
Comment (jglick) [inline test/src/test/java/jenkins/security/ClassFilterImplTest.java:162]: No longer a module, so no longer relevant to test.

**Review by jvz — APPROVED** [review]

File: war/pom.xml [hunk 3]
```diff
@@ -106,33 +106,59 @@ THE SOFTWARE.
       <version>5.13</version>
       <scope>test</scope>
     </dependency>
-    <dependency>
-      <groupId>org.jenkins-ci.modules</groupId>
-      <artifactId>instance-identity</artifactId>
-    </dependency>
     <dependency>
       <groupId>org.jenkins-ci.modules</groupId>
       <artifactId>ssh-cli-auth</artifactId>
```
Comment (jglick) [inline war/pom.xml:111]: Note that this (and `sshd`) pick up a transitive dep on `instance-identity 2.2` which then still gets packaged into the WAR as a module. Cannot fix until #5049 is merged (or this is merged into that).

File: core/src/main/resources/jenkins/split-plugin-cycles.txt [hunk 4]
```diff
@@ -32,3 +32,10 @@ junit jaxb
 bouncycastle-api jaxb
 command-launcher jaxb
 jdk-tool jaxb
+
+# JENKINS-55582
+bouncycastle-api instance-identity
+bouncycastle-api command-launcher
+trilead-api instance-identity
+jdk-tool instance-identity
+bouncycastle-api jdk-tool
```
[the remark this answers is not in the document]
Comment (timja) [reply core/src/main/resources/jenkins/split-plugin-cycles.txt:41]: yes, sshd PR worked fine through setup wizard.

Appears to be related to this PR

Also I get this on a clean install on startup:

```
2021-02-25 21:13:53.773+0000 [id=42]	WARNING	h.ExtensionFinder$GuiceFinder$SezpozModule#configure: Failed to load org.jenkinsci.modules.systemd_slave_installer.SlaveInstallerFactoryImpl
java.lang.ClassNotFoundException: org.jenkinsci.main.modules.instance_identity.InstanceIdentity
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:538)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
Caused: java.lang.NoClassDefFoundError: Lorg/jenkinsci/main/modules/instance_identity/InstanceIdentity;
	at java.base/java.lang.Class.getDeclaredFields0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredFields(Class.java:3061)
	at java.base/java.lang.Class.getDeclaredFields(Class.java:2248)
	at hudson.ExtensionFinder$GuiceFinder$SezpozModule.resolve(ExtensionFinder.java:497)
	at hudson.ExtensionFinder$GuiceFinder$SezpozModule.resolve(ExtensionFinder.java:482)
	at hudson.ExtensionFinder$GuiceFinder$SezpozModule.configure(ExtensionFinder.java:525)
	at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
	at com.google.inject.spi.Elements.getElements(Elements.java:110)
	at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:138)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:104)
	at com.google.inject.Guice.createInjector(Guice.java:96)
	at com.google.inject.Guice.createInjector(Guice.java:73)
	at hudson.ExtensionFinder$GuiceFinder.<init>(ExtensionFinder.java:285)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at net.java.sezpoz.IndexItem.instance(IndexItem.java:181)
	at hudson.ExtensionFinder$Sezpoz._find(ExtensionFinder.java:703)
	at hudson.ExtensionFinder$Sezpoz.find(ExtensionFinder.java:689)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:349)
	at hudson.ExtensionList.load(ExtensionList.java:381)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:317)
	at hudson.ExtensionList.getComponents(ExtensionList.java:183)
	at jenkins.model.Jenkins$6.onInitMilestoneAttained(Jenkins.java:1161)
	at jenkins.InitReactorRunner$1.onAttained(InitReactorRunner.java:84)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.lambda$onAttained$3(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.run(ReactorListener.java:109)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onAttained(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.Reactor$1.run(Reactor.java:177)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

[the remark this answers is not in the document]
Comment (jglick) [reply core/src/main/resources/jenkins/split-plugin-cycles.txt:41]: Will recheck after merge of #5049.

[the remark this answers is not in the document]
Comment (timja) [reply core/src/main/resources/jenkins/split-plugin-cycles.txt:41]: It’s merged?

[the remark this answers is not in the document]
Comment (jglick) [reply core/src/main/resources/jenkins/split-plugin-cycles.txt:41]: Yes I know, but I need time to retest.

**Comment by jglick** [comment]
Nope, cannot do this yet—need to do `slave-installer` first: https://github.com/jenkinsci/launchd-slave-installer-module/blob/1264c586db41b11cc1647dd3b126f695fee5c6de/src/main/java/org/jenkinsci/modules/launchd_slave_installer/SlaveInstallerFactoryImpl.java#L27 depends on `instance-identity`. Better suppress `instance-identity` from the update center.

Turns out doing them all together, as in #3988, was easier than trying to do one module at a time.

**Comment by jglick** [comment]
Now would like to merge `slave-installer` impls. Waiting to see if https://github.com/jenkinsci/windows-slave-installer-module/pull/41 gets merged.

**Comment by jglick** [comment]
Have not had time to work on this.

**Comment by basil** [comment]
Completed in #6570.

**Pull request closed without merging**